### PR TITLE
Subjects

### PR DESCRIPTION
**Pull request by sirageDb**

### CODE REVIEW

File: src/server.ts [hunk 1]
```diff
@@ -11,14 +11,15 @@ import { IConfig } from './config/env.dev';
...
+import SUbjectResolver from './controller/subjectResolver';
 
 export default async function startServer(
   config: IConfig,
 ): Promise<ApolloServer> {
   // on décrit un schéma graphQl à l'aide de la foncton buildSchema
   const schema: GraphQLSchema = await buildSchema({
     validate: false,
-    resolvers: [UserAuthResolver, ClassroomResolver, FlashcardResolver],
+    resolvers: [UserAuthResolver, ClassroomResolver, FlashcardResolver, SUbjectResolver],
```
Comment (jonathan-francois) [inline src/server.ts]:
```suggestion
    resolvers: [UserAuthResolver, ClassroomResolver, FlashcardResolver, SubjectResolver],
```

File: src/server.ts [hunk 2]
```diff
@@ -11,14 +11,15 @@ import { IConfig } from './config/env.dev';
 import UserAuthResolver from './controller/UserAuthResolver';
 import ClassroomResolver from './controller/ClassroomResolver';
 import FlashcardResolver from './controller/FlashcardResolver';
+import SUbjectResolver from './controller/subjectResolver';
```
Comment (jonathan-francois) [inline src/server.ts]:
```suggestion
import SubjectResolver from './controller/subjectResolver';
```

File: src/model/subject.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import mongoose from "mongoose";
+import { iSubject } from "../utils/types/classroomTypes";
+
+const subjects = new mongoose.Schema({
```
Comment (jonathan-francois) [inline src/model/subject.ts]:
```suggestion
const subjects = new mongoose.Schema<iSubject>({
```

File: src/model/graphql/subjectModelGQL.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
+import { Field, ObjectType, ID } from 'type-graphql';
+
```
Comment (jonathan-francois) [inline src/model/graphql/subjectModelGQL.ts:2]:
```suggestion

```

File: src/model/graphql/subjectModelGQL.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    _id! : string;
+
+    @Field()
+    imageUrl! : string;
+
+    @Field()
+    name! : string 
+}
+
+
```
Comment (jonathan-francois) [inline src/model/graphql/subjectModelGQL.ts]:
```suggestion

```

File: src/model/graphql/subjectModelGQL.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    @Field()
+    title!: string;
+  
+    @Field(() => [String])
+    tag!: string[];
+}
+
+@ObjectType()
+export class SubjectFlashcardModelGQL {
+    // returned subjectId
```
Comment (jonathan-francois) [inline src/model/graphql/subjectModelGQL.ts]:
```suggestion

```

File: src/controller/subjectResolver.ts [hunk 7]
```diff
@@ -0,0 +1,90 @@
+import {
+    Resolver,
+    Arg,
+    Query,
+} from 'type-graphql';
+
+import { ApolloError } from 'apollo-server-express';
+import { SubjectModelGQL, SubjectFlashcardModelGQL } from '../model/graphql/subjectModelGQL';
+import { iClassroom, iSubject } from '../utils/types/classroomTypes';
+import SubjectModel from "../model/subject";
```
Comment (jonathan-francois) [inline src/controller/subjectResolver.ts]:
```suggestion
import SubjectModel from '../model/subject';
```

File: src/controller/subjectResolver.ts [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+import ClassroomModel from "../model/classroom";
```
Comment (jonathan-francois) [inline src/controller/subjectResolver.ts]:
```suggestion
import ClassroomModel from '../model/classroom';
```

File: src/controller/subjectResolver.ts [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        }
+        catch {
+            throw new ApolloError("Cannot get subjects");
+        }
+    }
+
+
+    // private method to get subject located in a classroom by providing classroom object and subjectId
+    // =================================================
+    private getSubjectById(classroom: iClassroom, subjectId: string): iSubject | undefined {
```
Comment (jonathan-francois) [inline src/controller/subjectResolver.ts]:
```suggestion
    private getSubjectById(classroom: iClassroom, subjectId: string): iSubject | null{
```

File: src/controller/subjectResolver.ts [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+        const subject = classroom.subject.find(
+            (currentSubject: iSubject) => currentSubject.subjectId.toString() === subjectId
+        )
+        return subject
```
Comment (jonathan-francois) [inline src/controller/subjectResolver.ts]:
```suggestion
        return subject || null;
```

File: src/controller/subjectResolver.ts [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+        catch {
+            throw new ApolloError("cannot find classroom")
+        }
+    }
+
+    // =================================================
+    private async getSubjectsByIds(subjectsIds: string[]): Promise<iSubject[] | null> {
+        try {
+            const subjects = await SubjectModel.find({ _id: { $in: subjectsIds } });
+            return subjects;
```
Comment (jonathan-francois) [inline src/controller/subjectResolver.ts]:
```suggestion
            return subjects || null;
```

File: src/controller/subjectResolver.ts [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+
+@Resolver(SubjectModelGQL)
+export default class SubjectResolver {
+
+    // Private method to get a classroom by providing its id
+    // =================================================
+    private async getClassroomById(classroomId: string): Promise<iClassroom | null > {
+        try {
+            const classroom = await ClassroomModel.findOne({ _id: classroomId });
+            return classroom;
```
Comment (jonathan-francois) [inline src/controller/subjectResolver.ts]:
```suggestion
            return classroom || null;
```

File: src/controller/subjectResolver.ts [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+    // Get subjects flashcards 
+    // =================================================
+    @Query(() => SubjectFlashcardModelGQL)
+    public async getAllFlashcardsBySubject(
+        @Arg('classroomId') classroomId: string,
+        @Arg('subjectId') subjectId: string
+    ): Promise<iSubject | null> {
+
+        const classroom = await this.getClassroomById(classroomId);
+        const subject = classroom && this.getSubjectById(classroom, subjectId);
```
Comment (jonathan-francois) [inline src/controller/subjectResolver.ts]:
```suggestion
        if(!classroom){
            throw new ApolloError("Could not get classroom");
        }
        const subject = this.getSubjectById(classroom, subjectId);
```

**Review by jonathan-francois — CHANGES_REQUESTED** [review]

File: src/controller/subjectResolver.ts [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+    // Get classroom subjects by providing a classroomId
+    // =================================================
+    @Query(() => [SubjectModelGQL])
+    public async getAllSubjectsByClassroom(
+        @Arg('classroomId') classroomId: string
+    ): Promise<iSubject[] | null> {
+
+        try{
+            const classroom = await this.getClassroomById(classroomId);
+            const subjectsId: string[] = [];
```
Comment (jonathan-francois) [inline src/controller/subjectResolver.ts]: subjectInfo: { "subjectId": "_id"   }

File: src/controller/subjectResolver.ts [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+            }
+    
+            for (let i = 0; i < classroom.subject.length; i += 1) {
+                subjectsId.push(classroom.subject[i].subjectId);
+            }
+    
+            if (subjectsId.length === 0) {
+                throw new Error("No subjects in classroom");
+            }
+                const subjects = await this.getSubjectsByIds(subjectsId);
```
Comment (jonathan-francois) [inline src/controller/subjectResolver.ts]: Object.keys(subjectInfo)

File: src/controller/subjectResolver.ts [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+
+        try{
+            const classroom = await this.getClassroomById(classroomId);
+            const subjectsId: string[] = [];
+            if (!classroom) {
+                throw new Error("cannot find classroom");
+            }
+    
+            for (let i = 0; i < classroom.subject.length; i += 1) {
+                subjectsId.push(classroom.subject[i].subjectId);
```
Comment (jonathan-francois) [inline src/controller/subjectResolver.ts]: subjectInfo[classroom.subject[i].subjectId] = classroom.subject[i]._id

File: src/controller/subjectResolver.ts [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+            }
+    
+            if (subjectsId.length === 0) {
+                throw new Error("No subjects in classroom");
+            }
+                const subjects = await this.getSubjectsByIds(subjectsId);
+                return subjects;
```
Comment (jonathan-francois) [inline src/controller/subjectResolver.ts]:
```suggestion
                return subjects.map(e => e._id = subjectInfo[e._id]);
```

**Review by jonathan-francois — CHANGES_REQUESTED** [review]

**Review by jonathan-francois — DISMISSED** [review]

File: src/controller/ClassroomResolver.ts [hunk 18]
```diff
@@ -136,26 +136,36 @@ export default class ClassroomResolver {
...
-          student: { ...student, userId: student._id },
+          student: newStudent,
         },
       },
       { new: true },
     );
 
+    console.log(classRoom)
+
+
```
Comment (jonathan-francois) [inline src/controller/ClassroomResolver.ts]: a supprimer

File: src/controller/ClassroomResolver.ts [hunk 19]
```diff
@@ -136,26 +136,36 @@ export default class ClassroomResolver {
...
     if (classRoom) {
       await userModel.findOneAndUpdate(
-        { _id: student._id },
+        { _id: student.id },
```
Comment (jonathan-francois) [inline src/controller/ClassroomResolver.ts]: non laisse le _id, student provient de mongo directement, id n'existe pas c'est _id

**Review by jonathan-francois — APPROVED** [review]

**Review by JeanneSic — APPROVED** [review]

**Pull request merged**